sockeye/inference.py

```diff
@@ -1043,25 +1047,58 @@ def _beam_search(self,
             if self.restrict_lexicon:
                 best_word_indices[:] = vocab_slice_ids.take(best_word_indices)
 
-            # (4) get hypotheses and their properties for beam_size winning hypotheses (ascending)
+            # (4) Normalize the scores of newly finished hypotheses
+            newly_finished = (best_word_indices == self.vocab_target[C.EOS_SYMBOL]).expand_dims(axis=1)
+            scores_accumulated = mx.nd.where(newly_finished, scores_accumulated / self.length_penalty(lengths), scores_accumulated)
+
+            # (5) Prune out low-probability hypotheses
+            for sentno in range(self.batch_size):
+
+                # Set the active beam size to full. It will get reduced if items are filtered.
+                active_beam_size[sentno] = self.beam_size
```

Comment:
how does this work exactly? When hypothesis are removed because they're finished the active_beam_size will be reduced for the next step, but then be reset to beam_size for the step after that. Is that right?

Reply:
active_beam_size is initialized to 1 to handle the t=1 time step (this allows removing the first-step logic from before). Here, we set it to beam_size. If pruning occurs, it is set again to something smaller. This is the same as having an if...else block on pruning.